windows-installer/construct.py

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env python3
+import ctypes
+import subprocess
+import os
+import sys
+import shutil
+import time
+
+
+def main():
+    build_path, repo_path, venv_path, venv_python = _prepare_environment()
+
+    _copy_assets(build_path, repo_path)
+
+    installer_cfg_path = _generate_pynsist_config(repo_path, build_path)
+
+    _prepare_build_tools(venv_path, venv_python)
+    _compile_wheels(repo_path, build_path, venv_python)
+    _build_installer(installer_cfg_path, venv_path)
+
+    print('Done')
+
+
+def _build_installer(installer_cfg_path, venv_path):
+    print('Build the installer')
+    subprocess.check_call([os.path.join(venv_path, 'Scripts', 'pynsist.exe'), installer_cfg_path])
+
+
+def _compile_wheels(repo_path, build_path, venv_python):
+    print('Compile wheels')
+
+    wheels_path = os.path.join(build_path, 'wheels')
+    os.makedirs(wheels_path)
+
+    certbot_packages = ['acme', '.']
+    # Uncomment following line to include all DNS plugins in the installer
+    # certbot_packages.extend([name for name in os.listdir(repo_path) if name.startswith('certbot-dns-')])
+    wheels_project = [os.path.join(repo_path, package) for package in certbot_packages]
+
+    command = [venv_python, '-m', 'pip', 'wheel', '-w', wheels_path]
+    command.extend(wheels_project)
+    subprocess.check_call(command)
+
+
+def _prepare_build_tools(venv_path, venv_python):
+    print('Prepare build tools')
+    subprocess.check_call([sys.executable, '-m', 'venv', venv_path])
+    subprocess.check_call(['choco', 'upgrade', '-y', 'nsis'])
+    subprocess.check_call([venv_python, '-m', 'pip', 'install', '--upgrade', 'pip'])
+    subprocess.check_call([venv_python, '-m', 'pip', 'install', 'wheel', 'pynsist'])
+
+
+def _copy_assets(build_path, repo_path):
+    print('Copy assets')
+    if os.path.exists(build_path):
+        os.rename(build_path, '{0}.{1}.bak'.format(build_path, int(time.time())))
+    os.makedirs(build_path)
+    shutil.copy(os.path.join(repo_path, 'windows-installer', 'certbot.ico'), build_path)
+    shutil.copy(os.path.join(repo_path, 'windows-installer', 'run.bat'), build_path)
+
+
+def _generate_pynsist_config(repo_path, build_path):
+    print('Generate pynsist configuration')
+
+    installer_cfg_path = os.path.join(build_path, 'installer.cfg')
+
+    certbot_version = subprocess.check_output([sys.executable, '-c', 'import certbot; print(certbot.__version__)'],
+                                              universal_newlines=True, cwd=repo_path).strip()
+
+    with open(os.path.join(installer_cfg_path), 'w') as file_h:
+        file_h.write("""\
+[Application]
+name=Certbot
+version={certbot_version}
```

Comment:
Similarly to how you install DEB/RPM packages, having two versions of a program on Windows is really unusual. Most installers I know of will propose at least to uninstall the old version, or will juste overwrite it without question.

Do you have a use case in mind here ? Because for `certbot-auto` and OS packages, it is always one unique version at a time.

Reply:
No. I'm not interested in letting users install two copies of Certbot. I just wanted to verify that the overall approach here in general works for updates.

I brought up two copies of the software as an example of something someone writing installer software might want to allow in general, but it's not the behavior we'd like from a Certbot installer. When running a new version of the installer, we'd want to replace the existing installation rather than creating the new one.

I believe deb/rpm packages do this based on the package name so new versions of the package are recognized as updates to the existing one. I was curious how this worked here and if there was any documentation about it.

Reply:
No I could not find the details. I suppose that the metadata declared here https://nsis.sourceforge.io/Docs/Chapter4.html#versioninfo constitutes the identity of the program, and some internal logic is able to decide from it if another version is already installed;

After, the overwrite logic of the files is described here: https://nsis.sourceforge.io/Docs/Chapter4.html#flags (section 4.8.2.8)

Reply:
In response to https://github.com/certbot/certbot/pull/7324/files#r319290171:

I couldn't find the details either. I found things like https://github.com/takluyver/pynsist/issues/66 and tested it myself with the setup here and it seemed to work.

My last question related to this is if it is possible to run the NSIS installer non-interactively, possibly while setting various settings on the command line? I'm imagining while we'll want to install updates in this way and preserve settings like the installation location.

Reply:
Yes silent install is built-in. And you can specify the install dir from command line: https://nsis.sourceforge.io/Docs/Chapter3.html#installerusagecommon